File: Exercise25/Python/LinkList.py
import ListNode

class LinkList:
    def __init__(self) -> None:
        self.head = None
        self.tail = None
        self.size = 0
    
    def get(self, index) -> ListNode:
        assert(index >= 0 and index <= self.size)
        temp = self.head

        for i in range(1, index + 1):
            temp = temp.next
        
        return temp


    def insertNode(self, index, item):
        assert(index >= 0 and index <= self.size)

        if not isinstance(item, ListNode.ListNode):
            item = ListNode.ListNode(item)
        
        if index == 0:
            if self.tail == None:
                self.tail = item
            else:
                item.next = self.head
            self.head = item
        elif index == self.size:
            if self.head == None:
                self.head = item
            else:
                self.tail.next = item
            self.tail = item
        else:
            prevNode = self.get(index - 1)
            item.next = prevNode.next
            prevNode.next = item
        
        self.size += 1

    def delNode(self, index):
        assert(index >= 0 and index <= self.size)

        if index == 0:
            self.head = self.head.next
        elif index == self.size:
            self.tail = self.get(self.size)
            self.tail.next = None
        else:
            prevNode = self.get(index - 1)
            prevNode.next = prevNode.next.next
        
        self.size -= 1

    def travel(self):
        """遍历链表"""
        cur = self.head

        while cur != None:
            print(cur.data)
            cur = cur.next

File: Exercise25/JAVA/untitled/src/package2/LinkList.java
package package2;

public class LinkList<T> {

    //头结点
    public Node<T> head;
    //尾结点
    public Node<T> tail;
    //链表长度
    public int size;

    public LinkList(){
        this.head = new Node<>(null,null);
        size = 0;
    }

    public Node<T> get(int index) {
        if (index <0 || index >=this.size){
            return null;
        }else{
            Node<T> temp = this.head;
            for(int i =1; i <= index; i++){
                temp = temp.next;
            }
            return temp;
        }
    }

    public void insert(int index ,T data) throws Exception{
        if(index <0 ||index > this.size){
            throw new Exception("插入超出范围");
        }else{
            Node<T> newNode = new Node<>(data);
            //在头结点插入元素
            if (index ==0){
                if(this.size >0){
                    newNode.next = head;
                }
                head = newNode;
            }
            //在尾结点插入元素
            else if(index == this.size){
                Node<T> temp = tail;
                temp.next = newNode;
                this.tail  = newNode;

            }else{
                //在中间插入元素
                Node<T> preNode = get(index-1);
                newNode.next = preNode.next;
                preNode.next = newNode;

            }

        }
        this.size ++;

        if(size == 1){
            tail = head;
        }
    }

    public void del(int index )throws Exception{
        if (index <0 ||index >=this.size){
            throw new Exception("删除超出范围");
        }else{
            //删除头结点
            if (index == 0){
                this.head = this.head.next;

            }else if(index == this.size-1){    //删除尾结点
                Node<T> preNode = get(index -1);
                this.tail = preNode;
                preNode.next = null;
            }else{
                //删除中间结点
                Node<T> preNode = get(index -1);
                preNode.next = preNode.next.next;
            }

        }
        this.size --;

    }
}
